iree/compiler/Dialect/Flow/Transforms/DispatchLinalgOnTensors.cpp

```diff
@@ -326,48 +326,77 @@ bool isAlwaysClonedIntoDispatchOp(Operation *op) {
   return false;
 }
 
+/// Computes the values that will be eventually be used within the dispatch
+/// workgroup op but defined outside the op after all clonable operations are
+/// cloned into the region. Returns (by reference) the clonable operations too.
+void getUsedValuesDefinedAboveAfterCloningOps(
+    Region &region, OpBuilder &b, llvm::SetVector<Value> &valuesDefinedAbove,
+    llvm::SetVector<Operation *> &clonedOps) {
+  mlir::getUsedValuesDefinedAbove(region, valuesDefinedAbove);
+  OpBuilder::InsertionGuard g(b);
+  b.setInsertionPointToStart(&region.front());
+  llvm::SetVector<Operation *> tempClonedOps;
+  for (Value operand : valuesDefinedAbove) {
+    Operation *definingOp = operand.getDefiningOp();
+    if (!definingOp || !isAlwaysClonedIntoDispatchOp(definingOp)) continue;
+    clonedOps.insert(definingOp);
+    // Temporarily clone this operation. This is just so that when the values
```

Comment:
Ouch. IIUC, we are trying to only select the subset of values defined above from ops not gonna be copied into the region. What about using `mlir::visitUsedValuesDefinedAbove`, which has a callback to filter out values coming from `isAlwaysClonedIntoDispatchOp`?

Reply:
Well, I moved away from this. Updated the approach to use a worklist and backtracking.